hibernate-core/src/main/java/org/hibernate/query/sqm/tree/expression/ValueBindJpaCriteriaParameter.java

```diff
@@ -51,17 +55,33 @@ public void appendHqlString(StringBuilder hql, SqmRenderContext context) {
 	}
 
 	@Override
-	public boolean equals(Object o) {
-		return this == o;
+	// TODO: fix this
+	public int compareTo(SqmParameter<T> parameter) {
+		return this == parameter ? 0 : 1;
 	}
 
+	// this is not really a parameter, it's really a literal value
```

Comment:
Your implementation surely is better that the existing one, but this thing actually really is a parameter, in the sense that for caching, the value doesn't matter. This is a point where using `equals`/`hashCode` for cache keys is kind of quirky. Naturally, the equality of this type probably should check for value equality, but for caching, the only thing that matters IMO is the fact that the "other" object to compare against is also a parameter.

Reply:
Are you suggesting that a correct implementation is of `equals()` is:

```java
return object instanceof ValueBindJpaCriteriaParameter<?>
```

Reply:
Right, that's why I wrote this:

>This is a point where using `equals`/`hashCode` for cache keys is kind of quirky.

Essentially, this boils down to the "custom equals method with context parameter" suggestion I made in a different discussion on this PR.

Reply:
Yeah well I tried that and it was a bloodbath.

We could probably make the change you suggest and then fix whatever other code is making use of `equals()`. It could use `compareTo()` instead, for example.

Not going to do that in this PR though, since this seems to be working well enough for now. We can open an issue.